src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ContentDeliveryClient.cs

```diff
@@ -0,0 +1,371 @@
+                                            return PrepareListServicesAsync(innerCancellationToken2)
+                                                .WithUri(nextPage)
+                                                .Select(_ => _.Result.AsHttpApiCall());
+                                        };
+                                }
+
+                                ReadOnlyCollectionPage<Service> results = new BasicReadOnlyCollectionPage<Service>(list, prepareGetNextPageAsync);
+                                return results;
+                            });
+                };
+
+            return GetBaseUriAsync(cancellationToken)
+                .Then(PrepareRequestAsyncFunc(HttpMethod.Get, template, parameters, cancellationToken))
+                .Select(task => new ListServicesApiCall(CreateCustomApiCall(task.Result, HttpCompletionOption.ResponseContentRead, deserializeResult)));
+        }
+
+        /// <inheritdoc/>
+        public virtual Task<GetServiceApiCall> PrepareGetServiceAsync(ServiceId serviceId, CancellationToken cancellationToken)
+        {
+            if (serviceId == null)
+                throw new ArgumentNullException("serviceId");
+
+            UriTemplate template = new UriTemplate("services/{service_id}");
+            Dictionary<string, string> parameters = new Dictionary<string, string> { { "service_id", serviceId.Value } };
+            return GetBaseUriAsync(cancellationToken)
+                .Then(PrepareRequestAsyncFunc(HttpMethod.Get, template, parameters, cancellationToken))
+                .Select(task => new GetServiceApiCall(CreateJsonApiCall<Service>(task.Result)));
+        }
+
+        /// <inheritdoc/>
+        public virtual Task<AddServiceApiCall> PrepareAddServiceAsync(ServiceData serviceData, CancellationToken cancellationToken)
+        {
+            UriTemplate template = new UriTemplate("services");
+            Dictionary<string, string> parameters = new Dictionary<string, string>();
+
+            Func<HttpResponseMessage, CancellationToken, Task<ServiceId>> deserializeResult =
+                (responseMessage, innerCancellationToken) =>
+                {
+                    string absolutePath = responseMessage.Headers.Location.AbsolutePath;
+                    int lastSlash = absolutePath.LastIndexOf('/');
+                    return CompletedTask.FromResult(new ServiceId(absolutePath.Substring(lastSlash + 1)));
+                };
+
+            return GetBaseUriAsync(cancellationToken)
+                .Then(PrepareRequestAsyncFunc(HttpMethod.Post, template, parameters, serviceData, cancellationToken))
+                .Select(task => new AddServiceApiCall(CreateCustomApiCall(task.Result, HttpCompletionOption.ResponseContentRead, deserializeResult)));
+        }
+
+        /// <inheritdoc/>
+        public virtual Task<UpdateServiceApiCall> PrepareUpdateServiceAsync(ServiceId serviceId, ServiceData updatedServiceData, CancellationToken cancellationToken)
```

Comment:
This method deviates from one of the fundamental features of this library - API call preparation methods in a service interface prepare exactly one call and (aside from possibly authenticating) do not send any HTTP requests. The current implementation here combines two operations.

The standard pattern for implementing a combined call, such as this one or the Move Object operation that could appear in an Object Storage Service implementation, is to implement the operation as an extension method. I would expect to see the following:

1. A `PrepareUpdateServiceAsync` method here which takes a `PatchDocument` as an argument but does not do any calculations on it.
2. An `UpdateServiceAsync` method in `ContentDeliveryServiceExtensions` which works the way this method does (by requesting the current object and preparing the JSON patch).